src/frontend/js/data/SessionProvider/BaseSessionProvider.tsx

```diff
@@ -0,0 +1,94 @@
+import { PropsWithChildren, useCallback, useEffect, useMemo } from 'react';
+import { handle } from 'utils/errors/handle';
+import { AuthenticationApi } from 'utils/api/authentication';
+import { useQuery, useQueryClient } from 'react-query';
+import { Nullable } from 'types/utils';
+import { User } from 'types/User';
+import { REACT_QUERY_SETTINGS } from 'settings';
+import usePrevious from 'utils/usePrevious';
+import { Session } from './SessionContext';
+
+/**
+ * BaseSessionProvider
+ *
+ * @param children - Elements to render inside SessionProvider
+ *
+ * @return {Object} Session
+ * @return {Object} Session.user - authenticated user information
+ * @return {Function} Session.login - redirect to the login page
+ * @return {Function} Session.register - redirect to the register page
+ * @return {Function} Session.destroy - set Session to undefined then make a request to logout user to the authentication service
+ */
+const BaseSessionProvider = ({ children }: PropsWithChildren<any>) => {
+  /**
+   * `user` is:
+   * - `undefined` when we have not made the `whoami` request yet;
+   * - `null` when the user is anonymous or the request failed;
+   * - a user object when the user is logged in.
+   */
+  const { data: user } = useQuery<Nullable<User>>('user', AuthenticationApi!.me, {
+    refetchOnWindowFocus: true,
+    staleTime: REACT_QUERY_SETTINGS.staleTimes.session,
+  });
+  const previousUserState = usePrevious(user);
+
+  const queryClient = useQueryClient();
+
+  const login = useCallback(() => {
+    if (!AuthenticationApi) return handle(new Error('No AuthenticationApi configured!'));
```

Comment:
Line 29 you are sure `AuthenticationApi` is defined, probably you don't need this check

Reply:
Yes good catch !